File: examples_chirho/sudoku_chirho.py
#!/usr/bin/env python3
"""
Sudoku Solver using 1-Bit Domain Propagation ☧

Demonstrates practical use of miniKanren-style constraint propagation
with bit-parallel domain operations.

Each cell has a 9-bit domain where bit i means value (i+1) is possible.
Constraints (row, column, box) are applied via bitwise AND.

"Whether therefore ye eat, or drink, or whatsoever ye do, 
 do all to the glory of God." — 1 Corinthians 10:31
"""

import time
from typing import List, Optional, Tuple

# Domain constants
FULL_DOMAIN_CHIRHO = 0b111111111  # All values 1-9 possible
EMPTY_DOMAIN_CHIRHO = 0b000000000  # No values possible (failure)

def value_to_domain_chirho(v_chirho: int) -> int:
    """Convert value 1-9 to singleton domain."""
    if v_chirho < 1 or v_chirho > 9:
        raise ValueError(f"Value must be 1-9, got {v_chirho}")
    return 1 << (v_chirho - 1)

def domain_to_value_chirho(d_chirho: int) -> Optional[int]:
    """Convert singleton domain to value, or None if not singleton."""
    if d_chirho == 0 or (d_chirho & (d_chirho - 1)) != 0:
        return None  # Empty or multiple values
    return d_chirho.bit_length()

def popcount_chirho(d_chirho: int) -> int:
    """Count possible values in domain."""
    return bin(d_chirho).count('1')

def is_singleton_chirho(d_chirho: int) -> bool:
    """Check if exactly one value is possible."""
    return d_chirho != 0 and (d_chirho & (d_chirho - 1)) == 0

def lowest_value_chirho(d_chirho: int) -> Tuple[int, int]:
    """Fork: get (lowest single value, remaining values)."""
    lowest_chirho = d_chirho & (-d_chirho)  # Isolate lowest bit
    rest_chirho = d_chirho & (d_chirho - 1)  # Clear lowest bit
    return lowest_chirho, rest_chirho


class SudokuSolverChirho:
    """
    Sudoku solver using bit-parallel domain propagation.
    
    Each cell is a 9-bit domain. Propagation uses bitwise AND.
    Search uses branching on smallest domain (fail-first heuristic).
    """
    
    def __init__(self_chirho):
        # 81 cells, each with a 9-bit domain
        self_chirho.domains_chirho: List[int] = [FULL_DOMAIN_CHIRHO] * 81
        self_chirho.propagation_count_chirho = 0
        self_chirho.branch_count_chirho = 0
    
    def cell_index_chirho(self_chirho, row_chirho: int, col_chirho: int) -> int:
        """Convert (row, col) to flat index."""
        return row_chirho * 9 + col_chirho
    
    def get_peers_chirho(self_chirho, idx_chirho: int) -> List[int]:
        """Get indices of all cells that constrain this cell."""
        row_chirho = idx_chirho // 9
        col_chirho = idx_chirho % 9
        box_row_chirho = (row_chirho // 3) * 3
        box_col_chirho = (col_chirho // 3) * 3
        
        peers_chirho = set()
        
        # Same row
        for c_chirho in range(9):
            peers_chirho.add(row_chirho * 9 + c_chirho)
        
        # Same column
        for r_chirho in range(9):
            peers_chirho.add(r_chirho * 9 + col_chirho)
        
        # Same 3x3 box
        for r_chirho in range(box_row_chirho, box_row_chirho + 3):
            for c_chirho in range(box_col_chirho, box_col_chirho + 3):
                peers_chirho.add(r_chirho * 9 + c_chirho)
        
        peers_chirho.discard(idx_chirho)  # Remove self
        return list(peers_chirho)
    
    def set_value_chirho(self_chirho, idx_chirho: int, val_chirho: int) -> bool:
        """
        Set a cell to a specific value and propagate constraints.
        Returns False if this leads to failure.
        """
        domain_chirho = value_to_domain_chirho(val_chirho)
        
        # Intersect with current domain (unification!)
        self_chirho.domains_chirho[idx_chirho] &= domain_chirho
        
        if self_chirho.domains_chirho[idx_chirho] == EMPTY_DOMAIN_CHIRHO:
            return False
        
        return self_chirho.propagate_chirho()
    
    def propagate_chirho(self_chirho) -> bool:
        """
        Arc consistency propagation using bit operations.
        
        For each singleton domain, remove that value from all peers.
        Repeat until no changes (fixed point).
        """
        changed_chirho = True
        while changed_chirho:
            changed_chirho = False
            self_chirho.propagation_count_chirho += 1
            
            for idx_chirho in range(81):
                d_chirho = self_chirho.domains_chirho[idx_chirho]
                
                if d_chirho == EMPTY_DOMAIN_CHIRHO:
                    return False  # Failure
                
                if is_singleton_chirho(d_chirho):
                    # Remove this value from all peers
                    not_d_chirho = FULL_DOMAIN_CHIRHO ^ d_chirho
                    
                    for peer_idx_chirho in self_chirho.get_peers_chirho(idx_chirho):
                        old_chirho = self_chirho.domains_chirho[peer_idx_chirho]
                        # This is the key operation: domain intersection via AND
                        new_chirho = old_chirho & not_d_chirho
                        
                        if new_chirho != old_chirho:
                            changed_chirho = True
                            self_chirho.domains_chirho[peer_idx_chirho] = new_chirho
                            
                            if new_chirho == EMPTY_DOMAIN_CHIRHO:
                                return False  # Failure
        
        return True
    
    def is_solved_chirho(self_chirho) -> bool:
        """Check if all cells are singletons."""
        return all(is_singleton_chirho(d_chirho) for d_chirho in self_chirho.domains_chirho)
    
    def find_branch_cell_chirho(self_chirho) -> Optional[int]:
        """
        Find cell with smallest domain > 1 (fail-first heuristic).
        Returns None if all cells are singletons.
        """
        best_idx_chirho = None
        best_count_chirho = 10  # More than max possible
        
        for idx_chirho in range(81):
            count_chirho = popcount_chirho(self_chirho.domains_chirho[idx_chirho])
            if 1 < count_chirho < best_count_chirho:
                best_idx_chirho = idx_chirho
                best_count_chirho = count_chirho
        
        return best_idx_chirho
    
    def clone_chirho(self_chirho) -> 'SudokuSolverChirho':
        """Create a copy for branching."""
        new_chirho = SudokuSolverChirho()
        new_chirho.domains_chirho = self_chirho.domains_chirho.copy()
        new_chirho.propagation_count_chirho = self_chirho.propagation_count_chirho
        new_chirho.branch_count_chirho = self_chirho.branch_count_chirho
        return new_chirho
    
    def solve_chirho(self_chirho) -> bool:
        """
        Solve using constraint propagation + search.
        Returns True if solved, False if unsolvable.
        """
        if not self_chirho.propagate_chirho():
            return False
        
        if self_chirho.is_solved_chirho():
            return True
        
        # Branch on smallest domain
        branch_idx_chirho = self_chirho.find_branch_cell_chirho()
        if branch_idx_chirho is None:
            return self_chirho.is_solved_chirho()
        
        domain_chirho = self_chirho.domains_chirho[branch_idx_chirho]
        
        # Try each possible value
        while domain_chirho != 0:
            self_chirho.branch_count_chirho += 1
            
            # Fork: try lowest value first
            try_val_chirho, domain_chirho = lowest_value_chirho(domain_chirho)
            
            # Clone state and try this value
            attempt_chirho = self_chirho.clone_chirho()
            attempt_chirho.domains_chirho[branch_idx_chirho] = try_val_chirho
            
            if attempt_chirho.solve_chirho():
                # Success! Copy solution back
                self_chirho.domains_chirho = attempt_chirho.domains_chirho
                self_chirho.propagation_count_chirho = attempt_chirho.propagation_count_chirho
                self_chirho.branch_count_chirho = attempt_chirho.branch_count_chirho
                return True
        
        return False  # All branches failed
    
    def load_puzzle_chirho(self_chirho, puzzle_chirho: str) -> bool:
        """
        Load puzzle from string (81 chars, 0 or . for empty).
        Returns False if puzzle is invalid.
        """
        puzzle_chirho = puzzle_chirho.replace('\n', '').replace(' ', '')
        if len(puzzle_chirho) != 81:
            raise ValueError(f"Puzzle must be 81 chars, got {len(puzzle_chirho)}")
        
        for idx_chirho, ch_chirho in enumerate(puzzle_chirho):
            if ch_chirho in '123456789':
                if not self_chirho.set_value_chirho(idx_chirho, int(ch_chirho)):
                    return False
        
        return True
    
    def to_string_chirho(self_chirho) -> str:
        """Convert solution to string."""
        result_chirho = []
        for idx_chirho in range(81):
            val_chirho = domain_to_value_chirho(self_chirho.domains_chirho[idx_chirho])
            result_chirho.append(str(val_chirho) if val_chirho else '.')
        return ''.join(result_chirho)
    
    def print_board_chirho(self_chirho):
        """Pretty-print the board."""
        for row_chirho in range(9):
            if row_chirho % 3 == 0 and row_chirho > 0:
                print("------+-------+------")
            
            line_chirho = []
            for col_chirho in range(9):
                if col_chirho % 3 == 0 and col_chirho > 0:
                    line_chirho.append("|")
                
                idx_chirho = row_chirho * 9 + col_chirho
                val_chirho = domain_to_value_chirho(self_chirho.domains_chirho[idx_chirho])
                line_chirho.append(f" {val_chirho if val_chirho else '.'}")
            
            print(''.join(line_chirho))


def main_chirho():
    print("Sudoku Solver using 1-Bit Domain Propagation ☧\n")
    
    # Test puzzles (easy to hard)
    puzzles_chirho = [
        # Easy
        ("Easy",
         "530070000"
         "600195000"
         "098000060"
         "800060003"
         "400803001"
         "700020006"
         "060000280"
         "000419005"
         "000080079"),
        
        # Medium
        ("Medium",
         "000000680"
         "030000000"
         "900800100"
         "000002074"
         "060090030"
         "780500000"
         "001007005"
         "000000020"
         "024000000"),
        
        # Hard (17-clue minimal)
        ("Hard (17 clues)",
         "000000010"
         "400000000"
         "020000000"
         "000050407"
         "008000300"
         "001090000"
         "300400200"
         "050100000"
         "000806000"),
    ]
    
    for name_chirho, puzzle_chirho in puzzles_chirho:
        print(f"{'='*40}")
        print(f"Puzzle: {name_chirho}")
        print(f"{'='*40}")
        
        solver_chirho = SudokuSolverChirho()
        
        if not solver_chirho.load_puzzle_chirho(puzzle_chirho):
            print("Invalid puzzle!")
            continue
        
        print("\nInitial:")
        solver_chirho.print_board_chirho()
        
        start_chirho = time.perf_counter()
        solved_chirho = solver_chirho.solve_chirho()
        elapsed_chirho = (time.perf_counter() - start_chirho) * 1000
        
        if solved_chirho:
            print("\nSolved:")
            solver_chirho.print_board_chirho()
            print(f"\n✓ Solved in {elapsed_chirho:.2f}ms")
            print(f"  Propagation rounds: {solver_chirho.propagation_count_chirho}")
            print(f"  Search branches: {solver_chirho.branch_count_chirho}")
        else:
            print("\n✗ No solution exists")
        
        print()
    
    print("☧ Soli Deo Gloria ☧")


if __name__ == "__main__":
    main_chirho()

File: rust_chirho/src/sudoku_chirho.rs
//! Sudoku Solver using 1-Bit Domain Propagation ☧
//!
//! Demonstrates miniKanren-style constraint propagation with bit-parallel ops.
//! Each cell has a 9-bit domain where bit i means value (i+1) is possible.
//!
//! Techniques:
//! - **Naked singles**: Cell with one possibility → propagate to peers
//! - **Hidden singles**: Value in only one cell of unit → assign it
//! - **Naked pairs**: Two cells with same two values → eliminate from peers
//!
//! "Whether therefore ye eat, or drink, or whatsoever ye do,
//!  do all to the glory of God." — 1 Corinthians 10:31

// ============================================================================
// Constants
// ============================================================================

const GRID_SIZE_CHIRHO: usize = 9;
const NUM_CELLS_CHIRHO: usize = 81;
const FULL_DOMAIN_CHIRHO: u16 = 0b111111111;
const EMPTY_DOMAIN_CHIRHO: u16 = 0;

// ============================================================================
// Precomputed lookup tables (computed at compile time)
// ============================================================================

/// All 27 units: 9 rows + 9 columns + 9 boxes
static ALL_UNITS_CHIRHO: [[u8; 9]; 27] = compute_all_units_chirho();

/// Peer indices for each cell (20 peers per cell)
static PEERS_CHIRHO: [[u8; 20]; 81] = compute_peers_chirho();

const fn compute_all_units_chirho() -> [[u8; 9]; 27] {
    let mut units_chirho = [[0u8; 9]; 27];
    let mut unit_idx_chirho = 0usize;

    // Rows (units 0-8)
    let mut row_chirho = 0usize;
    while row_chirho < 9 {
        let mut col_chirho = 0usize;
        while col_chirho < 9 {
            units_chirho[unit_idx_chirho][col_chirho] = (row_chirho * 9 + col_chirho) as u8;
            col_chirho += 1;
        }
        unit_idx_chirho += 1;
        row_chirho += 1;
    }

    // Columns (units 9-17)
    let mut col_chirho = 0usize;
    while col_chirho < 9 {
        let mut row_chirho = 0usize;
        while row_chirho < 9 {
            units_chirho[unit_idx_chirho][row_chirho] = (row_chirho * 9 + col_chirho) as u8;
            row_chirho += 1;
        }
        unit_idx_chirho += 1;
        col_chirho += 1;
    }

    // Boxes (units 18-26)
    let mut box_chirho = 0usize;
    while box_chirho < 9 {
        let box_row_chirho = (box_chirho / 3) * 3;
        let box_col_chirho = (box_chirho % 3) * 3;
        let mut i_chirho = 0usize;
        let mut r_chirho = 0usize;
        while r_chirho < 3 {
            let mut c_chirho = 0usize;
            while c_chirho < 3 {
                units_chirho[unit_idx_chirho][i_chirho] =
                    ((box_row_chirho + r_chirho) * 9 + box_col_chirho + c_chirho) as u8;
                i_chirho += 1;
                c_chirho += 1;
            }
            r_chirho += 1;
        }
        unit_idx_chirho += 1;
        box_chirho += 1;
    }

    units_chirho
}

const fn compute_peers_chirho() -> [[u8; 20]; 81] {
    let mut result_chirho = [[0u8; 20]; 81];
    let mut idx_chirho = 0usize;

    while idx_chirho < 81 {
        let row_chirho = idx_chirho / 9;
        let col_chirho = idx_chirho % 9;
        let box_row_chirho = (row_chirho / 3) * 3;
        let box_col_chirho = (col_chirho / 3) * 3;
        let mut count_chirho = 0usize;

        // Row peers
        let mut c_chirho = 0usize;
        while c_chirho < 9 {
            let peer_chirho = row_chirho * 9 + c_chirho;
            if peer_chirho != idx_chirho {
                result_chirho[idx_chirho][count_chirho] = peer_chirho as u8;
                count_chirho += 1;
            }
            c_chirho += 1;
        }

        // Column peers
        let mut r_chirho = 0usize;
        while r_chirho < 9 {
            let peer_chirho = r_chirho * 9 + col_chirho;
            if peer_chirho != idx_chirho {
                result_chirho[idx_chirho][count_chirho] = peer_chirho as u8;
                count_chirho += 1;
            }
            r_chirho += 1;
        }

        // Box peers (excluding row/col overlap)
        let mut br_chirho = box_row_chirho;
        while br_chirho < box_row_chirho + 3 {
            let mut bc_chirho = box_col_chirho;
            while bc_chirho < box_col_chirho + 3 {
                let peer_chirho = br_chirho * 9 + bc_chirho;
                if peer_chirho != idx_chirho && br_chirho != row_chirho && bc_chirho != col_chirho {
                    result_chirho[idx_chirho][count_chirho] = peer_chirho as u8;
                    count_chirho += 1;
                }
                bc_chirho += 1;
            }
            br_chirho += 1;
        }

        idx_chirho += 1;
    }

    result_chirho
}

// ============================================================================
// Helper functions
// ============================================================================

#[inline]
const fn value_to_domain_chirho(v_chirho: u8) -> u16 {
    1u16 << (v_chirho - 1)
}

#[inline]
const fn domain_to_value_chirho(d_chirho: u16) -> u8 {
    if d_chirho == 0 || (d_chirho & (d_chirho - 1)) != 0 {
        0
    } else {
        d_chirho.trailing_zeros() as u8 + 1
    }
}

#[inline]
const fn is_singleton_chirho(d_chirho: u16) -> bool {
    d_chirho != 0 && (d_chirho & (d_chirho - 1)) == 0
}

#[inline]
const fn lowest_bit_chirho(d_chirho: u16) -> u16 {
    d_chirho & d_chirho.wrapping_neg()
}

// ============================================================================
// Solver
// ============================================================================

/// Sudoku solver using bit-parallel domain propagation
#[derive(Clone)]
pub struct SudokuSolverChirho {
    domains_chirho: [u16; NUM_CELLS_CHIRHO],
    pub propagation_count_chirho: u32,
    pub branch_count_chirho: u32,
}

impl Default for SudokuSolverChirho {
    fn default() -> Self {
        Self::new_chirho()
    }
}

impl SudokuSolverChirho {
    pub fn new_chirho() -> Self {
        Self {
            domains_chirho: [FULL_DOMAIN_CHIRHO; NUM_CELLS_CHIRHO],
            propagation_count_chirho: 0,
            branch_count_chirho: 0,
        }
    }

    // ========================================================================
    // Core propagation
    // ========================================================================

    /// Eliminate value from a cell's domain, return false if empty
    #[inline]
    fn eliminate_chirho(&mut self, idx_chirho: usize, val_bit_chirho: u16) -> bool {
        self.domains_chirho[idx_chirho] &= !val_bit_chirho;
        self.domains_chirho[idx_chirho] != EMPTY_DOMAIN_CHIRHO
    }

    /// Assign value to cell (set domain to singleton)
    #[inline]
    fn assign_chirho(&mut self, idx_chirho: usize, val_bit_chirho: u16) {
        self.domains_chirho[idx_chirho] = val_bit_chirho;
    }

    /// Propagate singleton constraints to peers
    fn propagate_naked_singles_chirho(&mut self) -> bool {
        for idx_chirho in 0..NUM_CELLS_CHIRHO {
            let d_chirho = self.domains_chirho[idx_chirho];

            if d_chirho == EMPTY_DOMAIN_CHIRHO {
                return false;
            }

            if is_singleton_chirho(d_chirho) {
                for &peer_chirho in &PEERS_CHIRHO[idx_chirho] {
                    if !self.eliminate_chirho(peer_chirho as usize, d_chirho) {
                        return false;
                    }
                }
            }
        }
        true
    }

    /// Find hidden singles in all units
    fn propagate_hidden_singles_chirho(&mut self) -> Result<bool, ()> {
        let mut changed_chirho = false;

        for unit_chirho in &ALL_UNITS_CHIRHO {
            for val_chirho in 0..GRID_SIZE_CHIRHO as u16 {
                let val_bit_chirho = 1u16 << val_chirho;

                // Count cells that can hold this value
                let (count_chirho, last_chirho) = self.count_value_in_unit_chirho(unit_chirho, val_bit_chirho);

                match count_chirho {
                    0 => return Err(()), // No cell can hold value = contradiction
                    1 if !is_singleton_chirho(self.domains_chirho[last_chirho]) => {
                        self.assign_chirho(last_chirho, val_bit_chirho);
                        changed_chirho = true;
                    }
                    _ => {}
                }
            }
        }

        Ok(changed_chirho)
    }

    /// Count cells in unit that can hold a value
    #[inline]
    fn count_value_in_unit_chirho(&self, unit_chirho: &[u8; 9], val_bit_chirho: u16) -> (u8, usize) {
        let mut count_chirho = 0u8;
        let mut last_chirho = 0usize;

        for &cell_chirho in unit_chirho {
            if self.domains_chirho[cell_chirho as usize] & val_bit_chirho != 0 {
                count_chirho += 1;
                last_chirho = cell_chirho as usize;
            }
        }

        (count_chirho, last_chirho)
    }

    /// Find and eliminate naked pairs
    fn propagate_naked_pairs_chirho(&mut self) -> bool {
        let mut changed_chirho = false;

        for unit_chirho in &ALL_UNITS_CHIRHO {
            changed_chirho |= self.find_naked_pairs_in_unit_chirho(unit_chirho);
        }

        changed_chirho
    }

    /// Find naked pairs within a single unit
    fn find_naked_pairs_in_unit_chirho(&mut self, unit_chirho: &[u8; 9]) -> bool {
        let mut changed_chirho = false;

        for i_chirho in 0..GRID_SIZE_CHIRHO {
            let cell_i_chirho = unit_chirho[i_chirho] as usize;
            let domain_i_chirho = self.domains_chirho[cell_i_chirho];

            if domain_i_chirho.count_ones() != 2 {
                continue;
            }

            for j_chirho in (i_chirho + 1)..GRID_SIZE_CHIRHO {
                let cell_j_chirho = unit_chirho[j_chirho] as usize;

                if self.domains_chirho[cell_j_chirho] == domain_i_chirho {
                    // Found naked pair - eliminate from other cells
                    for k_chirho in 0..GRID_SIZE_CHIRHO {
                        if k_chirho != i_chirho && k_chirho != j_chirho {
                            let cell_k_chirho = unit_chirho[k_chirho] as usize;
                            let old_chirho = self.domains_chirho[cell_k_chirho];
                            self.domains_chirho[cell_k_chirho] &= !domain_i_chirho;
                            changed_chirho |= self.domains_chirho[cell_k_chirho] != old_chirho;
                        }
                    }
                }
            }
        }

        changed_chirho
    }

    // ========================================================================
    // High-level propagation strategies
    // ========================================================================

    /// Basic propagation (naked singles only) - fast
    pub fn propagate_basic_chirho(&mut self) -> bool {
        let mut changed_chirho = true;
        while changed_chirho {
            changed_chirho = false;
            self.propagation_count_chirho += 1;

            for idx_chirho in 0..NUM_CELLS_CHIRHO {
                let d_chirho = self.domains_chirho[idx_chirho];

                if d_chirho == EMPTY_DOMAIN_CHIRHO {
                    return false;
                }

                if is_singleton_chirho(d_chirho) {
                    for &peer_chirho in &PEERS_CHIRHO[idx_chirho] {
                        let p_chirho = peer_chirho as usize;
                        let old_chirho = self.domains_chirho[p_chirho];
                        let new_chirho = old_chirho & !d_chirho;
                        if new_chirho != old_chirho {
                            self.domains_chirho[p_chirho] = new_chirho;
                            changed_chirho = true;
                            if new_chirho == EMPTY_DOMAIN_CHIRHO {
                                return false;
                            }
                        }
                    }
                }
            }
        }
        true
    }

    /// Full propagation (all techniques) - thorough
    pub fn propagate_chirho(&mut self) -> bool {
        loop {
            self.propagation_count_chirho += 1;

            if !self.propagate_naked_singles_chirho() {
                return false;
            }

            match self.propagate_hidden_singles_chirho() {
                Err(()) => return false,
                Ok(true) => continue,
                Ok(false) => {}
            }

            if self.propagate_naked_pairs_chirho() {
                continue;
            }

            break;
        }
        true
    }

    // ========================================================================
    // Solving
    // ========================================================================

    /// Find cell with smallest domain > 1 (MRV heuristic)
    fn find_branch_cell_chirho(&self) -> Option<usize> {
        let mut best_chirho: Option<(usize, u32)> = None;

        for (idx_chirho, &d_chirho) in self.domains_chirho.iter().enumerate() {
            let count_chirho = d_chirho.count_ones();
            if count_chirho > 1 {
                match best_chirho {
                    None => best_chirho = Some((idx_chirho, count_chirho)),
                    Some((_, best_count_chirho)) if count_chirho < best_count_chirho => {
                        best_chirho = Some((idx_chirho, count_chirho));
                        if count_chirho == 2 {
                            break; // Can't do better than 2
                        }
                    }
                    _ => {}
                }
            }
        }

        best_chirho.map(|(idx_chirho, _)| idx_chirho)
    }

    #[inline]
    pub fn is_solved_chirho(&self) -> bool {
        self.domains_chirho.iter().all(|&d_chirho| is_singleton_chirho(d_chirho))
    }

    fn count_unsolved_chirho(&self) -> usize {
        self.domains_chirho.iter().filter(|&&d_chirho| !is_singleton_chirho(d_chirho)).count()
    }

    /// Search with backtracking
    fn search_chirho(&mut self) -> bool {
        let branch_idx_chirho = match self.find_branch_cell_chirho() {
            Some(idx_chirho) => idx_chirho,
            None => return self.is_solved_chirho(),
        };

        // Save domains for backtracking (stack copy, no heap allocation)
        let saved_domains_chirho = self.domains_chirho;
        let mut domain_chirho = self.domains_chirho[branch_idx_chirho];

        while domain_chirho != 0 {
            self.branch_count_chirho += 1;

            let try_val_chirho = lowest_bit_chirho(domain_chirho);
            domain_chirho &= domain_chirho - 1;

            // Restore and try this value
            self.domains_chirho = saved_domains_chirho;
            self.domains_chirho[branch_idx_chirho] = try_val_chirho;

            if self.solve_adaptive_chirho() {
                return true;
            }
        }

        // Restore on complete failure
        self.domains_chirho = saved_domains_chirho;
        false
    }

    /// Adaptive solve - chooses strategy based on difficulty
    pub fn solve_adaptive_chirho(&mut self) -> bool {
        if !self.propagate_basic_chirho() {
            return false;
        }

        if self.is_solved_chirho() {
            return true;
        }

        // Hard puzzle? Use full propagation
        if self.count_unsolved_chirho() > 15 {
            if !self.propagate_chirho() {
                return false;
            }
            if self.is_solved_chirho() {
                return true;
            }
        }

        self.search_chirho()
    }

    /// Full solve (always uses advanced propagation)
    pub fn solve_chirho(&mut self) -> bool {
        if !self.propagate_chirho() {
            return false;
        }
        if self.is_solved_chirho() {
            return true;
        }
        self.search_chirho()
    }

    /// Fast solve (basic propagation only)
    pub fn solve_fast_chirho(&mut self) -> bool {
        if !self.propagate_basic_chirho() {
            return false;
        }
        if self.is_solved_chirho() {
            return true;
        }

        let branch_idx_chirho = match self.find_branch_cell_chirho() {
            Some(idx_chirho) => idx_chirho,
            None => return self.is_solved_chirho(),
        };

        // Save domains for backtracking
        let saved_domains_chirho = self.domains_chirho;
        let mut domain_chirho = self.domains_chirho[branch_idx_chirho];

        while domain_chirho != 0 {
            self.branch_count_chirho += 1;
            let try_val_chirho = lowest_bit_chirho(domain_chirho);
            domain_chirho &= domain_chirho - 1;

            self.domains_chirho = saved_domains_chirho;
            self.domains_chirho[branch_idx_chirho] = try_val_chirho;

            if self.solve_fast_chirho() {
                return true;
            }
        }

        self.domains_chirho = saved_domains_chirho;
        false
    }

    // ========================================================================
    // I/O
    // ========================================================================

    pub fn load_puzzle_chirho(&mut self, puzzle_chirho: &str) -> bool {
        let chars_chirho: Vec<char> = puzzle_chirho
            .chars()
            .filter(|c_chirho| !c_chirho.is_whitespace())
            .collect();

        if chars_chirho.len() != NUM_CELLS_CHIRHO {
            return false;
        }

        for (idx_chirho, ch_chirho) in chars_chirho.iter().enumerate() {
            if let Some(digit_chirho) = ch_chirho.to_digit(10) {
                if (1..=9).contains(&digit_chirho) {
                    self.domains_chirho[idx_chirho] = value_to_domain_chirho(digit_chirho as u8);
                }
            }
        }

        self.propagate_basic_chirho()
    }

    pub fn to_string_chirho(&self) -> String {
        self.domains_chirho
            .iter()
            .map(|&d_chirho| {
                let v_chirho = domain_to_value_chirho(d_chirho);
                if v_chirho > 0 { (b'0' + v_chirho) as char } else { '.' }
            })
            .collect()
    }

    #[inline]
    pub fn get_value_chirho(&self, idx_chirho: usize) -> u8 {
        domain_to_value_chirho(self.domains_chirho[idx_chirho])
    }

    #[inline]
    pub fn get_domain_chirho(&self, idx_chirho: usize) -> u16 {
        self.domains_chirho[idx_chirho]
    }
}

// ============================================================================
// Test puzzles
// ============================================================================

pub mod puzzles_chirho {
    pub const EASY_CHIRHO: &str = "530070000600195000098000060800060003400803001700020006060000280000419005000080079";
    pub const MEDIUM_CHIRHO: &str = "000000680030000000900800100000002074060090030780500000001007005000000020024000000";
    pub const HARD_CHIRHO: &str = "000000010400000000020000000000050407008000300001090000300400200050100000000806000";
    pub const ESCARGOT_CHIRHO: &str = "100007090030020008009600500005300900010080002600004000300000010040000007007000300";
    pub const INKALA_CHIRHO: &str = "800000000003600000070090200050007000000045700000100030001000068008500010090000400";
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests_chirho {
    use super::*;

    #[test]
    fn test_easy_chirho() {
        let mut s_chirho = SudokuSolverChirho::new_chirho();
        assert!(s_chirho.load_puzzle_chirho(puzzles_chirho::EASY_CHIRHO));
        assert!(s_chirho.solve_adaptive_chirho());
        assert!(s_chirho.is_solved_chirho());
    }

    #[test]
    fn test_hard_chirho() {
        let mut s_chirho = SudokuSolverChirho::new_chirho();
        assert!(s_chirho.load_puzzle_chirho(puzzles_chirho::HARD_CHIRHO));
        assert!(s_chirho.solve_adaptive_chirho());
        assert!(s_chirho.is_solved_chirho());
    }

    #[test]
    fn test_escargot_chirho() {
        let mut s_chirho = SudokuSolverChirho::new_chirho();
        assert!(s_chirho.load_puzzle_chirho(puzzles_chirho::ESCARGOT_CHIRHO));
        assert!(s_chirho.solve_adaptive_chirho());
        assert!(s_chirho.is_solved_chirho());
    }

    #[test]
    fn test_peers_chirho() {
        for idx_chirho in 0..NUM_CELLS_CHIRHO {
            let unique_chirho: std::collections::HashSet<_> = PEERS_CHIRHO[idx_chirho].iter().collect();
            assert_eq!(unique_chirho.len(), 20);
        }
    }

    #[test]
    fn test_domain_conversion_chirho() {
        for v_chirho in 1..=9u8 {
            let d_chirho = value_to_domain_chirho(v_chirho);
            assert!(is_singleton_chirho(d_chirho));
            assert_eq!(domain_to_value_chirho(d_chirho), v_chirho);
        }
    }
}
